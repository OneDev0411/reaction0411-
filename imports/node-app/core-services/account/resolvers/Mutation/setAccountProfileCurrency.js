import { decodeAccountOpaqueId } from "@reactioncommerce/reaction-graphql-xforms/account";

/**
 * @name Mutation/setAccountProfileCurrency
 * @method
 * @memberof Accounts/GraphQL
 * @summary resolver for the setAccountProfileCurrency GraphQL mutation
 * @param {Object} _ - unused
 * @param {Object} args.input - an object of all mutation arguments that were sent by the client
 * @param {String} [args.input.accountId] - The account ID, which defaults to the viewer account
 * @param {String} args.input.currencyCode - The currencyCode to add to user profile
 * @param {String} [args.input.clientMutationId] - An optional string identifying the mutation call
 * @param {Object} context - an object containing the per-request state
 * @returns {Object} setAccountProfileCurrency
 */
export default async function setAccountProfileCurrency(_, { input }, context) {
  const { accountId, currencyCode, clientMutationId = null } = input;
  const decodedAccountId = decodeAccountOpaqueId(accountId);

  const updatedAccount = await context.mutations.setAccountProfileCurrency(context, {
    currencyCode,
    accountId: decodedAccountId
  });

  return {
    account: updatedAccount,
    clientMutationId
  };
}
