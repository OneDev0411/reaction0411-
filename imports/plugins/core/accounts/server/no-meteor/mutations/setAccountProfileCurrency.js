import SimpleSchema from "simpl-schema";
import ReactionError from "@reactioncommerce/reaction-error";

const inputSchema = new SimpleSchema({
  currencyCode: String,
  decodedAccountId: String
});

/**
 * @name accounts/setAccountProfileCurrency
 * @memberof Mutations/Accounts
 * @summary Sets users profile currency
 * @param {Object} context - GraphQL execution context
 * @param {Object} input - Necessary input for mutation. See SimpleSchema.
 * @param {String} input.currencyCode - currency symbol to add to user profile
 * @param {String} [input.decodedAccountId] - optional decoded ID of account on which entry should be updated, for admin
 * @returns {Promise<Object>} with updated address
 */
export default async function setAccountProfileCurrency(context, input) {
  inputSchema.validate(input);
  const { appEvents, collections, userHasPermission, userId: userIdFromContext } = context;
  const { Accounts, Shops } = collections;
  const { currencyCode, decodedAccountId } = input;

  const accountId = decodedAccountId || userIdFromContext;
  if (!accountId) throw new ReactionError("access-denied", "You must be logged in to set profile currency");

  const account = await Accounts.findOne({ _id: accountId }, { fields: { shopId: 1 } });
  if (!account) throw new ReactionError("not-found", "No account found");

  if (userIdFromContext !== decodedAccountId) {
    if (!userHasPermission(["reaction-accounts"], account.shopId)) throw new ReactionError("access-denied", "Access denied");
  }

  // Make sure this currency code is in the related shop currencies list
  const shop = await Shops.findOne({ _id: account.shopId }, { fields: { currencies: 1 } });

  if (!shop || !shop.currencies || !shop.currencies[currencyCode]) {
    throw new ReactionError("invalid-argument", `The shop for this account does not define any currency with code "${currencyCode}"`);
  }

  const { value: updatedAccount } = await Accounts.findOneAndUpdate({ _id: accountId }, { $set: { "profile.currency": currencyCode } });

  await appEvents.emit("afterAccountUpdate", {
    account: updatedAccount,
    updatedBy: accountId,
    updatedFields: ["profile.currency"]
  });

  return updatedAccount;
}
