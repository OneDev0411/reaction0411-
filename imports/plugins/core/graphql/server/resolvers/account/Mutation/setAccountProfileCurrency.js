import { decodeAccountOpaqueId } from "@reactioncommerce/reaction-graphql-xforms/account";
import { getXformedCurrencyByCode } from "@reactioncommerce/reaction-graphql-xforms/currency";

/**
 * @name setAccountProfileCurrency
 * @method
 * @summary resolver for the setAccountProfileCurrency GraphQL mutation
 * @param {Object} _ - unused
 * @param {Object} args.input - an object of all mutation arguments that were sent by the client
 * @param {String} [args.input.accoundId] - The account ID, which defaults to the viewer account
 * @param {String} args.input.currencyCode - The currencyCode to add to user profile
 * @param {String} [args.input.clientMutationId] - An optional string identifying the mutation call
 * @param {Object} context - an object containing the per-request state
 * @return {Object} setAccountProfileCurrency
 */
export default function setAccountProfileCurrency(_, { input }, context) {
  const { accountId, currencyCode, clientMutationId = null } = input;
  const dbAccountId = decodeAccountOpaqueId(accountId);
  const updatedProfile = context.methods["accounts/setProfileCurrency"](context, [dbAccountId, currencyCode]);

  const userCurrency = getXformedCurrencyByCode(updatedProfile && updatedProfile.profile && updatedProfile.profile.currency);

  return {
    account: {
      currency: userCurrency,
      ...updatedProfile
    },
    clientMutationId
  };
}
