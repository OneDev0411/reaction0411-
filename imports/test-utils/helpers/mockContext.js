const mockContext = {
  accountId: "FAKE_ACCOUNT_ID",
  appEvents: {
    emit() {},
    on() {}
  },
  collections: {},
  getAbsoluteUrl: jest.fn().mockName("getAbsoluteUrl").mockImplementation((path) => {
    const adjustedPath = path[0] === "/" ? path : `/${path}`;
    return `https://app.mock${adjustedPath}`;
  }),
  getFunctionsOfType: jest.fn().mockName("getFunctionsOfType").mockReturnValue([]),
  mutations: {},
  queries: {},
  userHasPermission: jest.fn().mockName("userHasPermission"),
  userId: "FAKE_USER_ID"
};

/**
 * @summary Returns a mock collection instance with the given name
 * @param {String} collectionName The collection name
 * @returns {Object} Mock collection instance
 */
export function mockCollection(collectionName) {
  return {
    insert() {
      throw new Error("insert mongo method is deprecated, use insertOne or insertMany");
    },
    remove() {
      throw new Error("remove mongo method is deprecated, use deleteOne or deleteMany");
    },
    update() {
      throw new Error("update mongo method is deprecated, use updateOne or updateMany");
    },
    bulkWrite: jest.fn().mockName(`${collectionName}.bulkWrite`).mockReturnValue(Promise.resolve({
      nMatched: 2,
      nModified: 2,
      result: {
        writeErrors: []
      }
    })),
    deleteOne: jest.fn().mockName(`${collectionName}.deleteOne`).mockReturnValue(Promise.resolve({
      deletedCount: 1
    })),
    deleteMany: jest.fn().mockName(`${collectionName}.deleteMany`),
    find: jest
      .fn()
      .mockName(`${collectionName}.find`)
      .mockReturnThis(),
    findOne: jest.fn().mockName(`${collectionName}.findOne`),
    findOneAndDelete: jest.fn().mockName(`${collectionName}.findOneAndDelete`),
    findOneAndUpdate: jest.fn().mockName(`${collectionName}.findOneAndUpdate`),
    fetch: jest.fn().mockName(`${collectionName}.fetch`),
    insertOne: jest.fn().mockName(`${collectionName}.insertOne`),
    insertMany: jest.fn().mockName(`${collectionName}.insertMany`),
    toArray: jest.fn().mockName(`${collectionName}.toArray`),
    updateOne: jest.fn().mockName(`${collectionName}.updateOne`).mockReturnValue(Promise.resolve({
      matchedCount: 1,
      modifiedCount: 1
    })),
    updateMany: jest.fn().mockName(`${collectionName}.updateMany`)
  };
}

[
  "Accounts",
  "Assets",
  "Cart",
  "Catalog",
  "Emails",
  "ExampleIOUPaymentRefunds",
  "Groups",
  "MediaRecords",
  "NavigationItems",
  "NavigationTrees",
  "Notifications",
  "Orders",
  "Packages",
  "Products",
  "Revisions",
  "roles",
  "SellerShops",
  "Shipping",
  "Shops",
  "SimpleInventory",
  "Tags",
  "Templates",
  "Themes",
  "users"
].forEach((collectionName) => {
  mockContext.collections[collectionName] = mockCollection(collectionName);
});

mockContext.collections.Media = {
  find: jest.fn().mockName("Media.find"),
  findLocal: jest.fn().mockName("Media.findLocal"),
  findOne: jest.fn().mockName("Media.findOne"),
  findOneLocal: jest.fn().mockName("Media.findOneLocal"),
  insert: jest.fn().mockName("Media.insert"),
  update: jest.fn().mockName("Media.update"),
  remove: jest.fn().mockName("Media.remove")
};

export default mockContext;
