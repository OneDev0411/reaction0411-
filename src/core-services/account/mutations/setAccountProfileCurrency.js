import SimpleSchema from "simpl-schema";
import ReactionError from "@reactioncommerce/reaction-error";
import CurrencyDefinitions from "@reactioncommerce/api-utils/CurrencyDefinitions.js";

const inputSchema = new SimpleSchema({
  currencyCode: String,
  accountId: {
    type: String,
    optional: true
  }
});

/**
 * @name accounts/setAccountProfileCurrency
 * @memberof Mutations/Accounts
 * @summary Sets users profile currency
 * @param {Object} context - GraphQL execution context
 * @param {Object} input - Necessary input for mutation. See SimpleSchema.
 * @param {String} input.currencyCode - currency symbol to add to user profile
 * @param {String} [input.accountId] - optional decoded ID of account on which entry should be updated, for admin
 * @returns {Promise<Object>} with updated address
 */
export default async function setAccountProfileCurrency(context, input) {
  inputSchema.validate(input);
  const { appEvents, collections, accountId: accountIdFromContext } = context;
  const { Accounts } = collections;
  const { currencyCode, accountId: providedAccountId } = input;

  const accountId = providedAccountId || accountIdFromContext;
  if (!accountId) throw new ReactionError("access-denied", "You must be logged in to set profile currency");

  const account = await Accounts.findOne({ _id: accountId }, { projection: { shopId: 1 } });
  if (!account) throw new ReactionError("not-found", "No account found");

  await context.validatePermissions(`reaction:legacy:accounts:${account._id}`, "update:currency", {
    shopId: account.shopId,
    owner: account.userId,
    legacyRoles: ["reaction-accounts"]
  });

  if (!CurrencyDefinitions[currencyCode]) {
    throw new ReactionError("invalid-argument", `No currency has code "${currencyCode}"`);
  }

  const { value: updatedAccount } = await Accounts.findOneAndUpdate({
    _id: accountId
  }, {
    $set: { "profile.currency": currencyCode }
  }, {
    returnOriginal: false
  });

  await appEvents.emit("afterAccountUpdate", {
    account: updatedAccount,
    updatedBy: accountId,
    updatedFields: ["profile.currency"]
  });

  return updatedAccount;
}
