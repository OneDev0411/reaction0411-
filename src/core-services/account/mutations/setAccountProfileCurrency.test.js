import mockContext from "@reactioncommerce/api-utils/tests/mockContext.js";
import setAccountProfileCurrency from "./setAccountProfileCurrency.js";

test("throws if account not found - auth account", async () => {
  mockContext.collections.Accounts.findOne.mockReturnValueOnce(undefined);

  await expect(setAccountProfileCurrency({
    ...mockContext,
    accountId: "CONTEXT_ACCOUNT"
  }, { currencyCode: "USD" })).rejects.toThrowErrorMatchingSnapshot();

  expect(mockContext.collections.Accounts.findOne).toHaveBeenCalledWith({ _id: "CONTEXT_ACCOUNT" }, { projection: { shopId: 1 } });
});

test("throws if account not found - other account", async () => {
  mockContext.collections.Accounts.findOne.mockReturnValueOnce(undefined);

  await expect(setAccountProfileCurrency(mockContext, {
    accountId: "OTHER_ACCOUNT",
    currencyCode: "USD"
  })).rejects.toThrowErrorMatchingSnapshot();

  expect(mockContext.collections.Accounts.findOne).toHaveBeenCalledWith({ _id: "OTHER_ACCOUNT" }, { projection: { shopId: 1 } });
});

test("throws if currency code is invalid", async () => {
  mockContext.collections.Accounts.findOne.mockReturnValueOnce(Promise.resolve({ shopId: "SHOP_ID" }));

  await expect(setAccountProfileCurrency(mockContext, {
    accountId: "OTHER_ACCOUNT",
    currencyCode: "BAD_CODE"
  })).rejects.toThrowErrorMatchingSnapshot();
});

test("succeeds with valid currency code and account", async () => {
  mockContext.collections.Accounts.findOne.mockReturnValueOnce(Promise.resolve({ shopId: "SHOP_ID" }));
  mockContext.collections.Accounts.findOneAndUpdate.mockReturnValueOnce(Promise.resolve({ value: "MOCK_UPDATED_ACCOUNT" }));

  const updatedAccount = await setAccountProfileCurrency(mockContext, {
    accountId: "OTHER_ACCOUNT",
    currencyCode: "EUR"
  });

  expect(updatedAccount).toBe("MOCK_UPDATED_ACCOUNT");
  expect(mockContext.collections.Accounts.findOneAndUpdate).toHaveBeenCalled();
});
