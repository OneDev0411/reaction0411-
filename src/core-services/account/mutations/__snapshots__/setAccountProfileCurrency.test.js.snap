// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`throws if account not found - auth account 1`] = `"No account found"`;

exports[`throws if account not found - other account 1`] = `"No account found"`;

exports[`throws if currency code is invalid 1`] = `"No currency has code \\"BAD_CODE\\""`;
